lib/vfscore: Change faccessat with AT_SYMLINK_NOFOLLOW

robertZamfir601 opened the pull request:
* Change `faccessat` with `AT_SYMLINK_NOFOLLOW`.
* Move the path creation and logic before the `AT_SYMLINK_NOFOLLOW` (to avoid duplicate code)
* Apply `AT_SYMLINK_NOFOLLOW` only if the file pointed by the path is an actual symlink
* I didn't implement the `don't follow link logic`

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Change `faccessat` with `AT_SYMLINK_NOFOLLOW`:
* Move the path creation and logic before the `AT_SYMLINK_NOFOLLOW` (to avoid duplicate code)
* Apply `AT_SYMLINK_NOFOLLOW` only if the file pointed by the path is an actual symlink
* I didn't implement the `don't follow link logic`

StefanJum commented:
@robertZamfir601 does this work? Is there a reason why it's still a draft?

robertZamfir601 commented:
It is ready for review.

Mihnea0Firoiu reviewed (approved):
Looks good!
Reviewed-by: Mihnea Firoiu <mihneafiroiu0@gmail.com>

StefanJum reviewed (changes requested):
@robertZamfir601 just to make sure, the actual difference in functionality is that it won't crash if the target is not a symlink, right?

Looks fine, but please remove the `COMMIT_LOG_LONG_LINE` checkpatch ignore and fix the commit description, there is no need to keep the ling lines.

robertZamfir601 commented:
@StefanJum Yes, the only difference is that on a non-symlink target, the `AT_SYMLINK_NOFOLLOW` flag is ignored and the app won't crash.
I changed the description and removed `COMMIT_LOG_LONG_LINE` as well.

StefanJum reviewed (changes requested):
One more thing, please use your full name for the author and sign-off message (as in `Robert Zamfir <georobi.016@gmail.com>`). You are currently using `Robert` for both the author field and for the sign-off, and `Robert Zamfir` for the commiter.
Use `git config --global user.name`.

robertZamfir601 committed:
lib/vfscore: Change faccessat with AT_SYMLINK_NOFOLLOW

* Change faccessat with AT_SYMLINK_NOFOLLOW
* Move the path creation and logic before the AT_SYMLINK_NOFOLLOW
* Apply AT_SYMLINK_NOFOLLOW only if the file is an actual symlink
* I didn't implement the `don't follow link logic`

Checkpatch-Ignore: LONG_LINE_STRING
Checkpatch-Ignore: STRLCPY
Signed-off-by: Robert Zamfir <georobi.016@gmail.com>

robertZamfir601 commented:
Fixed the names

StefanJum reviewed (approved):
All good, thanks.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

andreittr commented on lib/vfscore/main.c:
```diff
@@ -2061,37 +2061,44 @@ UK_SYSCALL_R_DEFINE(int, access, const char*, pathname, int, mode)
-	}
+	if (pathname[0] == '/' ||
+	    dirfd == AT_FDCWD) {	/* Full path inside pathname */
+		strlcpy(p, pathname, PATH_MAX);
```
One minor comment: your change causes `faccessat` to always copy the path to a stack temporary even if it's a full (absolute) path that you could pass on to other functions as-is.
A more economical solution would be something like
```c
char tmp[PATH_MAX];
const char *p;

if (pathname[0] == '/' || dirfd == AT_FDCWD) {
    p = pathname;
} else {
    /* strcat the path together into tmp */
    p = tmp;
}
```
then use `p` as usual.
Or, if you're feeling particularly clever:
```c
const int fullpath = pathname[0] == '/' || dirfd == AT_FDCWD;
char tmp[fullpath ? PATH_MAX : 1];
const char *p;

if (fullpath) {
    p = pathname;
} else {
    /* build path in tmp */
    p = tmp;
}
```
in order to save stack space when not needing `tmp`.

robertZamfir601 replied:
First Issue: You are right. I changed the code exactly as you said.
Second Part: I moved the declaration of `tmp` inside the `else` branch. The compiler should be clever with it. I hope it is ok.

andreittr replied:
IIRC it is undefined behavior (or at the very least _very_ non-standard-C) to leak the address of a local variable outside of its scope -- that's explicitly why I didn't propose declaring `tmp` in the else clause, because you need it in the top-level function block.

If the 2nd option looks too hacky declaring `tmp` with a runtime size, you can leave it always at `PATH_MAX` and rely on the compiler being clever enough to not actually do a `sub %rsp` unless `tmp` is used. That at least will be well-defined behavior.

robertZamfir601 replied:
I understand. I moved the declaration back to the top-level, with `PATH_MAX` size.

andreittr replied:
Changes look good, thanks! Could you please squash them into the first commit and force-push the PR branch? Then I can give approval and we can merge.

As a policy we try to keep the commit history clean and reflecting self-contained changes, rather than tracking every adjustment during review -- so force-pushes are the common way to address comments.

robertZamfir601 replied:
Squashed everything together. Thank you!

andreittr reviewed (approved):
Approved-by: Andrei Tatar <andrei@unikraft.io>
